src/ORM/EagerLoadedDataList.php

```diff
@@ -0,0 +1,893 @@
+{
+    /**
+     * List responsible for instantiating the actual DataObject objects from eager-loaded data
+     */
+    private DataList $dataList;
+
+    /**
+     * Underlying DataObject class for this list
+     */
+    private string $dataClass;
+
+    /**
+     * The ID(s) of the record that owns this list if the list represents a relation
+     * Used for aggregations
+     */
+    private int|array|null $foreignID;
+
+    /**
+     * ID-indexed array that holds the data from SQL queries for the list
+     * @var array<int,array>
+     */
+    private array $rows = [];
+
+    /**
+     * Nested eager-loaded data which applies to relations on records contained in this list
+     * @var array<int,self|DataObject>
+     */
+    private array $eagerLoadedData = [];
+
+    private array $extraFields = [];
+
+    private array $limitOffset = [null, 0];
+
+    private string|array $sort = [];
+
+    public function __construct(string $dataListClass, string $dataClass, int|array|null $foreignID = null, array $manyManyComponent = [])
+    {
+        if (!is_a($dataListClass, DataList::class, true)) {
+            throw new LogicException('$dataListClass must be an instanceof DataList');
+        }
+
+        // relation lists require a valid foreignID or set of IDs
+        if (is_a($dataListClass, RelationList::class, true) && !$this->isValidForeignID($foreignID)) {
+            throw new InvalidArgumentException('$foreignID must be a valid ID for eager loaded relation lists');
+        }
+
+        $this->dataClass = $dataClass;
+        $this->foreignID = $foreignID;
+
+        // many_many relation lists have extra constructor args that don't apply for has_many or non-relations
+        if (is_a($dataListClass, ManyManyThroughList::class, true)) {
+            $this->dataList = ManyManyThroughList::create(
+                $dataClass,
+                // If someone instantiates one of these and passes DataObjectSchema::manyManyComponent() directly
+                // the class will be in here as 'join'
+                $manyManyComponent['joinClass'] ?? $manyManyComponent['join'],
+                $manyManyComponent['childField'],
+                $manyManyComponent['parentField'],
+                $manyManyComponent['extraFields'],
+                $dataClass,
+                $manyManyComponent['parentClass']
+            );
+        } elseif (is_a($dataListClass, ManyManyList::class, true)) {
+            $this->dataList = ManyManyList::create(
+                $dataClass,
+                $manyManyComponent['join'],
+                $manyManyComponent['childField'],
+                $manyManyComponent['parentField'],
+                $manyManyComponent['extraFields']
+            );
+        } else {
+            $this->dataList = $dataListClass::create($dataClass, '');
+        }
+
+        if (isset($manyManyComponent['extraFields'])) {
+            $this->extraFields = $manyManyComponent['extraFields'];
+        }
+    }
+
```

Comment:
Missing a few methods from DataList that aren't here - though depending on whether we want to keep the word `DataList` in the class name will probably change what methods we want:

- addFilter
- addMany
- createSearchFilter
- newObject
- removeAll
- removeByFilter
- removeByID
- removeMany
- subtract (at very least want this given we have add(), sum() etc)

Reply:
IMO None of these make sense in the context of eager-loaded data:
- removeAll
- removeByFilter
- removeByID
- removeMany
These are all above _removing_ data from the _database_ in the context of `DataList` and its subclasses.
In the case of `DataList`, all of these remove the actual records from the database.
In the case of the various relation lists, they remove the _relation_ between two records.
I think it is a much clearer experience for these methods to simply _not exist_ on this list, so calling them gives you an exception, and therefore you are aware that you were trying to do something this list doesn't support.

`addMany` doesn't make sense to implement here for a similar reason to the remove methods not being sensible - we don't add `DataObject` records, we add rows of data. So that method would already _not work the same_ as it does on `DataList`. I think it's better to just not implement it to make it clear that's not supported here.

Similarly, `newObject` doesn't make sense - the PHPDoc says "Return a new item to add to this DataList" but again we don't add `DataObject` records. It's a read-only list.

`createSearchFilter` is protected and therefore don't need to be implemented here

I will implement `addFilter` and `Subtract` for the sake of compatability.